src/main/scala/rise/core/Expr.scala

```diff
@@ -1,46 +1,82 @@
 package rise.core
 
+import semantics._
 import rise.core.types._
 import rise.core.DSL.Type.TypeEqual
 import rise.core.ShowRise._
 
+object alphaEquiv {
+  type TYPEEQ = Type => Type => Boolean
+
+  val equiv : TYPEEQ => Expr => Expr => Boolean = typeEq => a => b => {
+    typeEq(a.t)(b.t) && ((a, b) match {
+      case (Identifier(na), Identifier(nb)) => na == nb
+      case (Literal(da), Literal(db)) => da == db
+      case (a : Primitive, b : Primitive) => a.name == b.name
+      // Application compares structurally
+      case (App(fa, ea), App(fb, eb)) => equiv(typeEq)(fa)(fb) && equiv(typeEq)(ea)(eb)
+      case (DepApp(fa, xa), DepApp(fb, xb)) => equiv(typeEq)(fa)(fb) && xa == xb
+      // Abstraction compares after substitution
+      case (Lambda(xa, ta), other@Lambda(xb, _)) =>
+        typeEq(xa.t)(xb.t) && equiv(typeEq)(ta)(typedLifting.liftFunExpr(other).value(xa))
+      case (DepLambda(xa, ea), other@DepLambda(xb, _)) => (xa, xb) match {
+        case (n: NatIdentifier, _: NatIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[NatKind](other).value(n))
+        case (dt: DataTypeIdentifier, _: DataTypeIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[DataKind](other).value(dt))
+        case (addr: AddressSpaceIdentifier, _: AddressSpaceIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[AddressSpaceKind](other).value(addr))
+        case (n2n: NatToNatIdentifier, _: NatToNatIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[NatToNatKind](other).value(n2n))
+        case (n2d: NatToDataIdentifier, _: NatToDataIdentifier) =>
+          equiv(typeEq)(ea)(typedLifting.liftDepFunExpr[NatToDataKind](other).value(n2d))
+        case _ => false
+      }
+      case _ => false
+    })
+  }
+
+  val hash : Expr => Int = {
+    case _: Identifier => 17
```

Comment:
maybe in the future we can use some sort of DeBruijn indices to get a more useful hash, e.g. it would differentiate `\x. \y. x y` and `\x. \y. y x`. not sure if that it worth the effort but maybe we can add a comment about this in the code.

Reply:
As in, implement an on the fly conversion to de Bruijn? Maybe we could switch altogether (first rise and then DPIA), but yeah, it's worth documenting properly what this hash differentiates and what it doesn't.

Reply:
Yeah, on-the-fly with a simple `Map[Name, Int]` where you start with index 0 and increase when you go into lambdas, although this might be too expensive for a hash, not sure. I am not convinced that we want to use DeBruijn indices in the entire codebase, I am more tempted by graph representations, but that is another question ;p.